python/beeai_framework/tools/tool.py

```diff
@@ -115,25 +135,32 @@ def _create_emitter(self) -> Emitter:
         pass
 
     @abstractmethod
-    async def _run(self, input: IN, options: OPT | None, context: RunContext) -> OUT:
+    async def _run(self, input: TInput, options: TRunOptions | None, context: RunContext) -> TOutput:
         pass
 
-    def validate_input(self, input: IN | dict[str, Any]) -> IN:
+    def validate_input(self, input: TInput | dict[str, Any]) -> TInput:
         try:
             return self.input_schema.model_validate(input)
         except ValidationError as e:
             raise ToolInputValidationError("Tool input validation error", cause=e)
 
-    def run(self, input: IN | dict[str, Any], options: OPT | None = None) -> Run[OUT]:
-        async def run_tool(context: RunContext) -> OUT:
+    def run(self, input: TInput | dict[str, Any], options: TRunOptions | None = None) -> Run[TOutput]:
+        async def run_tool(context: RunContext) -> TOutput:
             error_propagated = False
+            context.emitter.events = {
+                "start": ToolStartEvent[TInput],
+                "success": ToolSuccessEvent[TInput],
+                "error": ToolErrorEvent[TInput],
+                "retry": ToolErrorEvent[TInput],
+                "finish": NoneType,
+            }
```

Comment:
This does not seem right. You can set default events in base `_create_emitter` method, but user should be able to extend them.

Reply:
I will tackle this in the followup PR